### PR TITLE
[SPARK-44945][DOCS][PYTHON] Automate PySpark error class documentation

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR introduces a new script named `errors_doc_gen.py` that automatically generates an RST documentation file containing the error classes defined in `pyspark.errors.error_classes`.

### Why are the changes needed?

With the increasing number of error classes in PySpark, it becomes imperative to have a centralized documentation that provides an overview of these errors for both developers and users. This automated script ensures that the documentation is always up-to-date with the error classes defined in the source code, reducing the manual effort required to update the documentation each time a new error class is introduced or an existing one is modified.

### Does this PR introduce _any_ user-facing change?

No, this script is primarily for internal use by developers to generate documentation. End-users will benefit from the updated documentation that this script produces, but they will not directly interact with the script itself.

### How was this patch tested?

The script was executed locally, and the generated RST file was verified for accuracy and completeness. Furthermore, the generated RST was rendered using Sphinx to ensure it produces the expected visual output.

Screenshot 2023-08-24 at 10 19 52 PM

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: python/docs/source/development/errors.rst [hunk 1]
```diff
@@ -1,92 +0,0 @@
-..  Licensed to the Apache Software Foundation (ASF) under one
```
Comment (itholic) [inline python/docs/source/development/errors.rst:1]: The `python/docs/source/development/errors.rst` file was removed as it's now dynamically generated during documentation build. This ensures that the error documentation is always up-to-date with the source code.

**Comment by bjornjorgensen** [comment]
HINT 
it is quite easy to see that you are using AI tools on this.
```
            f.write("-" * len(error_key) + "\n\n")  # Underline for the error key
            messages = error_details["message"]
            for message in messages:
                # Escape parentheses with a backslash when they follow a backtick
                message = re.sub(r"`(\()", r"`\\\1", message)
                f.write(message + "\n")
            f.write("\n")  # Add an empty line for better readability
```

I think you have to answer Yes to this one
Was this patch authored or co-authored using generative AI tooling?

**Comment by itholic** [comment]
> I think you have to answer Yes to this one
Was this patch authored or co-authored using generative AI tooling?

Actually I asked the generative AI to suggest a comment, but this script is just a slight modification of the code I wrote a long while ago :-)

That is, excluding the two comments, `# Underline for the error key` and `# Add an empty line for better readability`, the rest of the entire code was not written with the assistance of AI tooling at all.

I wonder if we can still consider AI as a co-author in such cases? If so, I would simply remove the two comments added by the AI instead of answer "Yes" since it seems not make enough sense to me.

cc @zero323 @HyukjinKwon Can I happen to ask your thought on this? This seems like a bit of a subtle issue

**Comment by itholic** [comment]
> I think you have to answer Yes to this one
Was this patch authored or co-authored using generative AI tooling?

Actually I asked the generative AI to suggest a comment, but basically this script is just a slight modification of the code I wrote a long while ago :-)

That is, excluding the two comments, `# Underline for the error key` and `# Add an empty line for better readability`, the rest of the entire code was not written with the assistance of AI tooling at all.

I wonder if we can still consider AI as a co-author in such cases? If so, I would simply remove the two comments suggested by the AI instead of answer "Yes" since it seems not make enough sense to me.

cc @zero323 @HyukjinKwon Can I happen to ask your thought on this? This seems like a bit of a subtle issue

**Comment by itholic** [comment]
I see the documentation build is succeed on CI:
```
copying static files... ... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build succeeded.

The HTML pages are in build/html.
```

**Comment by HyukjinKwon** [comment]
I think it's fine. In fact it's not the part of the actual code (and not even part of the release)

**Comment by HyukjinKwon** [comment]
You can even just rewrite the comment on your own @itholic

**Comment by itholic** [comment]
Updated 2 comments with self-written comments:
```diff
- # Underline for the error key
+ # The length of the error class name and underline must be the same
+ # to satisfy the RST format.
```

```diff
- # Add an empty line for better readability
+ # Add 2 new lines between the descriptions of each error class
+ # to improve the readability of the generated RST file.
```

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**